### ISSUE TITLE
Proper evaluation order of `if` operator in readtags expressions?

### ISSUE THREAD

**Issue by AmaiKinono** [universal-ctags/ctags#2817]
The `if` operator has a syntax like:

```
(if cond exp-true exp-false)
```

From what I've seen, I think readtags eval both `exp-true` and `exp-false` first, then pick one based on `cond`. For example:

``` console
$ readtags -Q '(if $extras #t #f)' -l
```

This lists tags with `extras:` field, which is expected. Then I can change it to this:

``` console
$ readtags -Q '(if $extras (print $extras) #f)' -l
```

I get output like this:

```
...
#f
#f
"inputFile"
citre-util.el   citre-util.el   1
#f
#f
#f
#f
"inputFile"
citre.el        citre.el        1
...
```

Notice that missing values are printed (as `#f`). If the condition is evaluated first, then either `exp-true` or `exp-false` is evaluated, this shouldn't happen.

I want the "right" evaluation order because string operators and `<>` don't work on missing values (#f), so `if` is used to rule out missing values. For filters expressions, I have this workaround:

``` lisp
change this:
(if (#/some-regexp/ $extras) some-expr #f)
to this:
(and $extras (#/some-regexp/ $extras) some-expr)
```

But for sorter expressions, I need this:

``` lisp
(if (and $extras &extras)
    (<> (if (#/regexp/ $extras) -1 1)
        (if (#/regexp/ &extras) -1 1))
  ;; For tags without extras: field, the order is uncertain.
  0)
```

**Comment by masatake** [comment]
Thank you for reporting this.
I took a poor mistake.

```
iff --git a/dsl/dsl.c b/dsl/dsl.c
index 4def0786d..9a87f4baf 100644
--- a/dsl/dsl.c
+++ b/dsl/dsl.c
@@ -125,7 +125,7 @@ static DSLProcBind pbinds [] = {
 	  .helpstr = "(and exp0 ... expN) -> <boolean>" },
 	{ "or",      sform_or,     NULL, DSL_PATTR_SELF_EVAL,
 	  .helpstr = "(or exp0 ... expN) -> <boolean>" },
-	{ "if",      sform_if,       NULL, DSL_PATTR_CHECK_ARITY, 3,
+	{ "if",      sform_if,       NULL, DSL_PATTR_SELF_EVAL|DSL_PATTR_CHECK_ARITY, 3,
 	  .helpstr = "(if cond exp-true exp-false) -> exp-true|exp-false" },
 	{ "not",     builtin_not,    NULL, DSL_PATTR_CHECK_ARITY, 1,
 	  .helpstr = "(not exp) -> <boolean>" },
```

I'll make a pull request.

**Comment by masatake** [comment]
I must fix #2759 first.
